File: src/main/kotlin/com/example/home_recipe/domain/refrigerator/Refrigerator.java
package com.example.home_recipe.domain.refrigerator;

import jakarta.persistence.UniqueConstraint;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;

import java.util.HashSet;

@Entity
public class Refrigerator {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id", nullable = false, unique = true)
    private Long userId;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "fridges_ingredients",
            joinColumns = @JoinColumn(name = "refrigerator_id"),
            inverseJoinColumns = @JoinColumn(name = "ingredient_id"),
            uniqueConstraints = @UniqueConstraint(
                    name = "uk_fridge_ingredient", columnNames = {"refrigerator_id", "ingredient_id"}
            )
    )
    private Set<Ingredient> ingredients = new HashSet<>();
}

File: src/main/kotlin/com/example/home_recipe/domain/refrigerator/Refrigerator.kt
package com.example.home_recipe.domain.refrigerator

import jakarta.persistence.*
import com.example.home_recipe.domain.ingredient.Ingredient

@Entity
@Table(name = "refrigerator")
class Refrigerator protected constructor() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "fridges_ingredients",
        joinColumns = [JoinColumn(name = "refrigerator_id")],
        inverseJoinColumns = [JoinColumn(name = "ingredient_id")],
        uniqueConstraints = [UniqueConstraint(
            name = "uk_fridge_ingredient",
            columnNames = ["refrigerator_id", "ingredient_id"]
        )]
    )
    val ingredients: MutableSet<Ingredient> = mutableSetOf()

    companion object {
        fun create(defaultIngredients: Collection<Ingredient> = emptyList()): Refrigerator =
            Refrigerator().apply {
                defaultIngredients.forEach(::addIngredient)
            }
    }

    fun addIngredient(ingredient: Ingredient): Boolean =
        ingredients.add(ingredient)

    fun useIngredient(ingredient: Ingredient): Boolean =
        ingredients.remove(ingredient)

    fun removeIngredientById(ingredientId: Long): Boolean {
        val target = ingredients.firstOrNull { it.id == ingredientId } ?: return false
        return ingredients.remove(target)
    }

    fun ingredientNames(): List<String> =
        ingredients.map { it.name }

    fun hasIngredient(ingredient: Ingredient): Boolean =
        ingredient in ingredients
}
